File: 二叉树/最大二叉树.go
package 二叉树

//这题秒杀了

func constructMaximumBinaryTree(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}
	maxIndex := 0
	maxNum := nums[0]
	for i := 1; i < len(nums); i++ { //找到数组最大值的下标
		if nums[i] > maxNum {
			maxNum = nums[i]
			maxIndex = i
		}
	}
	root := &TreeNode{nums[maxIndex], nil, nil}
	root.Left = constructMaximumBinaryTree(nums[:maxIndex])
	root.Right = constructMaximumBinaryTree(nums[maxIndex+1:])
	return root
}

File: 二叉树/最大二叉树.java
package 二叉树;

public class 最大二叉树 {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        return help(nums,0,nums.length-1);
    }

    public TreeNode help(int[] nums, int start, int end) {
        if (start > end) {
            return null;
        }
        int maxIndex = start;
        for (int i = start + 1; i <= end; i++) {
            if (nums[i] > nums[maxIndex]) {
                maxIndex = i;
            }
        }
        TreeNode root = new TreeNode(nums[maxIndex]);
        root.left = help(nums, start, maxIndex - 1);
        root.right = help(nums, maxIndex + 1, end);
        return root;
    }
}
